Fix platform-specific substitutions

sbomer opened the pull request:
These feature switches are dead code on non-windows. They were causing
linker warnings because the substitutions weren't being resolved.

See https://github.com/dotnet/sdk/pull/17676

sbomer committed:
These feature switches are dead code on non-windows. They were causing
linker warnings because the substitutions weren't being resolved.

See https://github.com/dotnet/sdk/pull/17676

eerhardt commented:
Do we know why these warnings aren't being emitted during [illink-sharedframework.targets](https://github.com/dotnet/runtime/blob/main/src/libraries/illink-sharedframework.targets) ?

cc @joperezr

vitek-karas commented on src/coreclr/System.Private.CoreLib/src/Internal/Runtime/InteropServices/InMemoryAssemblyLoader.cs:
```diff
@@ -12,9 +12,11 @@ namespace Internal.Runtime.InteropServices
     /// </summary>
     public static class InMemoryAssemblyLoader
     {
+#if TARGET_WINDOWS
```
Technically the whole class can be removed from the product on non-Windows - the IJW is not supported on non-Windows. @jkoritzinsky should be able to confirm this.

jkoritzinsky replied:
Yes, IJW is unsupported on non-Windows platforms.

sbomer replied:
I think this is being kept around so that this test can build on all platforms: https://github.com/dotnet/runtime/blob/main/src/tests/Interop/IJW/LoadIjwFromModuleHandle/LoadIjwFromModuleHandle.cs#L43

vitek-karas replied:
That's a bit suspect, since that will always fail. But I'm fine keeping it if @jkoritzinsky thinks it's worth it.

jkoritzinsky replied:
It's because in CI we build all the managed tests once on Linux, so the API I think needs to exist for the test to build.

marek-safar replied:
Do you build your test agast SPC.dll, that's quite unusual.

sbomer replied:
Yeah, it builds against SPC:https://github.com/dotnet/runtime/blob/fe8ae98d813788605ab034449ec7d313e8baa1ef/src/tests/Interop/IJW/LoadIjwFromModuleHandle/LoadIjwFromModuleHandle.csproj#L4

sbomer commented:
@eerhardt I think it's just because we don't set any feature settings during shared framework trim, so the linker won't try to apply them.

joperezr commented:
>Is it because we are running in library mode?

AFAIK the only thing that library mode does is that it will trim rooting only public surface area of the library in question, whereas non-library mode will root everything else, but @sbomer can confirm. I think in this case, as @sbomer  pointed out the reason why we don't get them is because we don't set any feature settings when we manually call the linker.

eerhardt commented:
> we don't set any feature settings during shared framework trim, so the linker won't try to apply them.

Does the validation of the method/type happen only when that feature value is set? Or when the linker tries parsing the document?

The reason I'm drilling into this is because I think it would be better if we caught problems like this in `dotnet/runtime`, and not upstack in the `dotnet/sdk` tests. So I'd hope we could use our illink-sharedframework target to verify these kinds of warnings don't happen.

vitek-karas commented:
Currently if the linker sees a feature switch and it evaluates the condition to false, it will simply skip the entire element. So it doesn't even read the XML in question.
It's a bit tricky to say that everything in the XML should be resolvable all the time. We currently don't use feature switches to "hide" things which are not even there, but somebody might want to do that...

sbomer committed:
Fix condition

Check OS, not Platform. Use the same pattern as we do for other
Windows-specific XMLs.

eerhardt reviewed (approved):
LGTM. Thanks for fixing this up.
